### PR TITLE
Switch vendored etcd to v3.5.6

### PR DESCRIPTION
**Pull request by dims**
We updated etcd server version to v3.5.6 in https://github.com/kubernetes/kubernetes/pull/114093

In this PR, we update the vendored dependency as well.

Related to https://github.com/kubernetes/kubernetes/issues/104007 as well, we got rid of one more  redundant dependency `form3tech-oss/jwt-go`

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
ONE more down!!

```
@@ -1,6 +1,6 @@
 {
-	"directDependencies": 212,
-	"transitiveDependencies": 234,
-	"totalDependencies": 285,
+	"directDependencies": 211,
+	"transitiveDependencies": 233,
+	"totalDependencies": 284,
 	"maxDepthOfDependencies": 26
 }
\ No newline at end of file
```

File: vendor/go.etcd.io/etcd/client/v3/watch.go [hunk 1]
```diff
@@ -688,6 +718,11 @@ func (w *watchGrpcStream) run() {
 	}
 }
 
+func shouldRetryWatch(cancelReasonError error) bool {
+	return (strings.Compare(cancelReasonError.Error(), v3rpc.ErrGRPCInvalidAuthToken.Error()) == 0) ||
+		(strings.Compare(cancelReasonError.Error(), v3rpc.ErrGRPCAuthOldRevision.Error()) == 0)
```
Comment (liggitt) [inline vendor/go.etcd.io/etcd/client/v3/watch.go:723]: this is stringifying the parameter error twice when it doesn't need to, and is stringifying constant API errors twice when they will never change

it's also unclear that exact string comparisons are correct for these... is there zero chance they will get wrapped by other errors that will add prefixes or suffixes? I'm surprised this isn't checking errors in a typed way

File: vendor/go.etcd.io/etcd/client/v3/watch.go [hunk 2]
```diff
@@ -580,6 +589,26 @@ func (w *watchGrpcStream) run() {
 
 			switch {
 			case pbresp.Created:
+				cancelReasonError := v3rpc.Error(errors.New(pbresp.CancelReason))
+				if shouldRetryWatch(cancelReasonError) {
```
Comment (liggitt) [inline vendor/go.etcd.io/etcd/client/v3/watch.go:593]: this looks really weird... is this handling a canceled flow in the created case? is that correct?

even if that's correct, it looks like this is triggering even in non-error cases (pbresp.CancelReason == "") - it's always creating a new error, a new rpc error, and calling `shouldRetryWatch` (I also commented in that method that stringification is happening more than it should)

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 1 above)
Comment (dims) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: cc @ahrtr @serathius

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 2 above)
Comment (dims) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:593]: cc @ahrtr @serathius

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 1 above)
Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: > this is stringifying the parameter error twice


Indeed there is no need to stringify the error twice. I just delivered a PR to improve this: https://github.com/etcd-io/etcd/pull/14935

Thanks @liggitt for pointing this out.




> it's also unclear that exact string comparisons are correct for these.

It's correct here. etcd defines a set of [gRPC errors](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/v3rpc/rpctypes/error.go#L24-L93) on server side. We need to translate each error to a string before responding to the client side. 

The logic here is that we need to retry on some error types (`ErrGRPCInvalidAuthToken` and `ErrGRPCAuthOldRevision`). So we need to compare the error message. Note that we have strict error messages definition, there is no any wrap & unwrap.

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 2 above)
Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:593]: See my comment above https://github.com/kubernetes/kubernetes/pull/114403#discussion_r1046397696

**Comment by ahrtr** [comment]
/lgtm
/approve

Thanks @dims

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 1 above)
Comment (liggitt) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: > etcd defines a set of [gRPC errors](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/v3rpc/rpctypes/error.go#L24-L93) on server side. We need to translate each error to a string before responding to the client side.

I'm really surprised the client is expected to compare error message strings to drive behavior. Isn't that fragile with skewed client and server versions?

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 2 above)
Comment (liggitt) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:593]: if pbresp.CancelReason is empty, shouldn't we avoid all of this added code?

Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:593]: Resolved.

File: vendor/go.etcd.io/etcd/client/v3/watch.go (hunk 1 above)
Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: > I'm really surprised the client is expected to compare error message strings to drive behavior. Isn't that fragile with skewed client and server versions?

This is a valid concern. This should be a technical debt. We may consider to define client side [codes](https://github.com/grpc/grpc-go/blob/eeb9afa1f6b6388152955eeca8926e36ca94c768/codes/codes.go#L29) in future.

Comment (liggitt) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: This change is expanding the etcd API contract... the client now requires specific error strings from the server in order to behave correctly. Kubernetes describes working against a storage layer that implements the etcd wire protocol, but this change is expanding the etcd API contract in unstructured and undocumented ways.

I don't really think this is a great change in general, but I *really* don't think etcd should add this without explicit documentation of the exact message strings required to drive this behavior, and tests ensuring the server doesn't change in the future in ways that will break this.

cc @lavalamp @deads2k for visibility

Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: cc @mitake and @ptabor as well

Probably we should add a field something like `reasonCode` in [WatchResponse](https://github.com/etcd-io/etcd/blob/ee9db729da87bd7a694909b5b9a2ab816f55125c/api/etcdserverpb/rpc.pb.go#L2192-L2221)



> Kubernetes describes working against a storage layer that implements the etcd wire protocol

Do you mean Kubernetes doesn't use the etcd `clientv3` at some places where it needs to talk to etcdserver at all?

Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: FYI. https://github.com/etcd-io/etcd/discussions/14992

Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: @dims Yes, will do it for 3.5 today.

Comment (ahrtr) [reply vendor/go.etcd.io/etcd/client/v3/watch.go:723]: FYI. https://github.com/etcd-io/etcd/pull/14995

**Comment by liggitt** [comment]
/test pull-kubernetes-dependencies
to run check added in https://github.com/kubernetes/kubernetes/pull/114952

**Pull request closed without merging**